keps/sig-node/3857-rro-hostpath/README.md

```diff
@@ -0,0 +1,909 @@
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+Support recursively read-only mounts for runc >= 1.1 && kernel >= 5.12.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+Support recursively read-only mounts for old runc and old kernel releases.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants to mount `/mnt`, includings its submounts such as `/mnt/usbstorage`, as read-only.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+Constraints: needs runc >= 1.1 && kernel >= 5.12.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Risk: a manifest with this feature will not work on old distros (See "Constraints" above).
+
+Mitigation: create a separate manifest for old distros.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### Core API
+Add `RecursivelyReadOnly *bool` to the [`VolumeMount`](https://github.com/kubernetes/kubernetes/blob/v1.26.1/pkg/apis/core/types.go#L1854-L1880) struct.
+
+When `RecursivelyReadOnly` is set to true, `ReadOnly` must be true as well.
+
+A pod manifest will look like this:
+```yaml
+spec:
+  volumes:
+    - name: foo
+      hostPath: 
+        path: /mnt
+        type: Directory
+  containers:
+  - volumeMounts:
+    - mountPath: /mnt
+      name: foo
+      readOnly: true
+      # NEW
+      recursivelyReadOnly: true
```

Comment:
> What workloads may be relying on submounts not being read-only today? 

- Mount `/src` as RO, while leaving `/src/_output` writable
- Mount `/lecture_materials` as RO, while leaving `/lecture_materials/student_assignment_submission` writable

> Should this behavior be folded into existing readOnly?

That will cause confusion as a Pod manifest specifically written for kernel >= 5.12 can be deployable on kernel < 5.12 nodes without causing an error.
Introducing a new `recursivelyReadOnly` property prohibits this by raising `ENOSYS` on kernel < 5.12.

Reply:
I think the question is "should this just be automatic when supported" ?  I am surprised to learn it wasn't ALWAYS this way. 

That said, this is a breaking change, and as such we should be really careful.  Our choices are:

1) Make a breaking change and risk users jobs failing.
2) Make this an "opt-in to security" situation.

Less API is better, so I wish we didn't need this, but it seems we do.  We really don't want to do breaking changes.

That said, I think we need to be really clear on the semantics:

The `readOnly` field defaults to false.  What does it mean if I set `readOnly` to false and `recursiveReadOnly` to true?  Is that allowed?  Do we change the default for `readOnly` to be "false unless recursiveReadOnly is true" ?  Or is it just an API error?

What happens if the user sets this and it lands on a node which doesn't support it?   If we ignore errors (== ignore this field_, we are ignoring a security directive, but if we fail the pod, there's no easy way for a user to know if their pod will EVER be able to run.  Could use some input from sig-security  @tabbysable @IanColdwater 

Do we want this as yet another bool or should we make it an enum?

Reply:
> 1. Make a breaking change and risk users jobs failing.
> 2. Make this an "opt-in to security" situation.

Purposely chose 1, to avoid giving users a false sense of security.
CrashLoopBackOffs are sometimes more preferable than potential data falsification.

> What does it mean if I set `readOnly` to false and `recursiveReadOnly` to true? Is that allowed? 

Disallowed (explained in the KEP).

> What happens if the user sets this and it lands on a node which doesn't support it?

runc fails with `ENOSYS`.

> if we fail the pod, there's no easy way for a user to know if their pod will EVER be able to run.

Kubelet may check the kernel version, and return a human-friendly error message that would be visible in `k get pod -o yaml`, etc.

> Do we want this as yet another bool or should we make it an enum?

What would be enum values?

Reply:
My idea for enums https://github.com/kubernetes/enhancements/pull/3858#discussion_r1100869332

Reply:
> Purposely chose 1, to avoid giving users a false sense of security.

The breaking change I meant would be to silently convert `readOnly` -> into RRO when possible and warning when not possible.  The result would be "my app (which depended on non-RO submounts) worked on Monday, but on Tueday my app fails, all because my cluster provider did a node upgrade".  As niche as that is, it's still a break.

I *think* you are proposing that people who opt-in to RRO would fail to run if RRO was not available.  Is that right?  The problem, of course, is that you have to convince people to take action to defend against a scenario that doesn't cause them measurable pain *and* that opt-in comes with risk.  That's sort of the worst option.

I assert that our tolerance for silently breaking people, even for niche security issues is approximately zero.  I even brought it to sig-arch today and while it was a small audience, the main sentiment was "we don't break people".

So the goal here is to be as secure as possible for as many people as possible while not silently breaking anyone.  I'll comment more on your other thread

Reply:
Updated the PR to take a ternary value: `recursivelyReadOnly: (Disabled|IfPossible|Required)`